### PR TITLE
Add a countdown timer to contest entry and voting pages

### PR DESCRIPTION
**Pull request by nekodex**
Should help visualise contest deadlines, cuz timezones are hard or something.

### CODE REVIEW

File: resources/assets/coffee/shared.coffee [hunk 1]
```diff
@@ -117,3 +117,8 @@ $(document).on 'keydown', '.js-quick-submit', (e) ->
 
   e.preventDefault()
   $(e.target).closest('form').submit()
+
+# countdown timers
+$(document).on 'turbolinks:load', ->
+  reactTurbolinks.register 'countdownTimer', CountdownTimer, (e) ->
```
Comment (nanaya) [inline resources/assets/coffee/shared.coffee]: move to main and without the `turbolinks:load` part.

File: resources/assets/coffee/react/_components/countdown-timer.coffee [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    divs = []
+    _.each fields, (value, field) ->
+      divs.push div key: field, className: "#{bn}__field",
+        div className: "#{bn}__digit",
+          if value < 10 then "0#{value}" else value
+        div className: "#{bn}__label", field
+
+    div className: bn,
+      div className: "#{bn}__header", "#{osu.trans('common.time.remaining')}:"
+      divs
```
Comment (nanaya) [inline resources/assets/coffee/react/_components/countdown-timer.coffee]:
```
for own field, value of fields
  div className: 'x',
    value
```

File: resources/assets/coffee/react/_components/countdown-timer.coffee [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    clearInterval @timer if diff == 0
+
+    @setState
+      diff: diff
+
+  render: =>
+    diff = @diff(@state.deadline) / 1000
+
+    fields =
+      'days': Math.floor(diff / (60 * 60 * 24))
```
Comment (nanaya) [inline resources/assets/coffee/react/_components/countdown-timer.coffee]: no quotes

File: resources/assets/coffee/react/_components/countdown-timer.coffee [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      'hours': Math.floor((diff / (60 * 60)) % 24)
+      'minutes': Math.floor((diff / 60) % 60)
+      'seconds': Math.floor(diff % 60)
+
+    divs = []
+    _.each fields, (value, field) ->
+      divs.push div key: field, className: "#{bn}__field",
+        div className: "#{bn}__digit",
+          if value < 10 then "0#{value}" else value
```
Comment (nanaya) [inline resources/assets/coffee/react/_components/countdown-timer.coffee]: `npm install left-pad`

...or just `_.padStart value, 2, 0` (not sure if worth)

File: resources/assets/coffee/react/_components/countdown-timer.coffee [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+      diff: @diff(@props.deadline)
+
+  componentDidMount: ->
+    @timer = setInterval @updateTimer, 1000
+
+  componentWillUnmount: ->
+    clearInterval @timer
+
+  diff: (time) ->
+    diff = moment.utc(time).diff()
```
Comment (nanaya) [inline resources/assets/coffee/react/_components/countdown-timer.coffee]: utc not needed?

File: resources/assets/coffee/react/_components/countdown-timer.coffee [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+{div} = React.DOM
+el = React.createElement
+
+bn = 'countdown-timer'
+
+class @CountdownTimer extends React.Component
+  constructor: (props) ->
+    super props
+    @state =
+      deadline: @props.deadline
```
Comment (nanaya) [inline resources/assets/coffee/react/_components/countdown-timer.coffee]: looks like parsing it once (`moment(@props.deadline)`) and calling the diff repeatedly is enough.

File: resources/assets/coffee/react/_components/countdown-timer.coffee [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+# You should have received a copy of the GNU Affero General Public License
+# along with osu!web.  If not, see <http://www.gnu.org/licenses/>.
+###
+{div} = React.DOM
+el = React.createElement
+
+bn = 'countdown-timer'
+
+class @CountdownTimer extends React.Component
+  constructor: (props) ->
```
Comment (nanaya) [inline resources/assets/coffee/react/_components/countdown-timer.coffee]: \~Slightly\~ shorter.
```
class Stuff ...
  componentWillMount: =>
    @deadline = moment(@props.deadline)

  componentWillReceiveProps: =>
    @deadline = moment(@props.deadline)

  componentWillUnmount: =>
    Timeout.clear @timer

  render: =>
    diff = @deadline.diff() / 1000
    Timeout.set(1000, @forceUpdate) if diff > 0

    fields = ...
    div
      ... 
```

File: resources/assets/coffee/shared.coffee [hunk 8]
```diff
@@ -117,3 +117,8 @@ $(document).on 'keydown', '.js-quick-submit', (e) ->
...
+    deadline: $(e).data('deadline')
```
Comment (nanaya) [inline resources/assets/coffee/shared.coffee]: `e.dataset.deadline` also works.

File: resources/assets/coffee/react/_components/countdown-timer.coffee (hunk 4 above)
Comment (nekodex) [inline resources/assets/coffee/react/_components/countdown-timer.coffee]: ಠ╭╮ಠ

Comment (nanaya) [inline resources/assets/coffee/react/_components/countdown-timer.coffee]: ಠ◡ಠ

**Pull request merged**